src/number_theory/kummer_dedekind.lean

```diff
@@ -0,0 +1,328 @@
+import ring_theory.ideal.operations
+import ring_theory.polynomial.basic
+import ring_theory.power_basis
+import ring_theory.unique_factorization_domain
+import data.polynomial.field_division
+import ring_theory.chain_of_divisors
+import algebra.algebra.basic
+
+/-!
+# Kummer-Dedekind theorem
+
+This file proves the Kummer-Dedekind theorem on the splitting of prime ideals in an extension
+of the ring of integers.
+
+## Main definitions
+
+ * `factors_equiv'` : The bijection in the Kummer-Dedekind theorem
+
+## Main results
+
+ * `normalized_factors_ideal_map_eq_normalized_factors_min_poly_mk_map` : The Kummer-Dedekind
+    theorem
+ * `ideal.irreducible_map_of_irreducible_minpoly` : `I.map (algebra_map R S)` is irreducible if
+    `(map I^.quotient.mk (minpoly R pb.gen))` is irreducible, where `pb` is a power basis of `S`
+    over `R`
+
+## TODO
+ * Define the conductor ideal and prove the Kummer-Dedekind theorem in full generality
+ * Prove the converse of `ideal.irreducible_map_of_irreducible_minpoly`
+
+## Tags
+
+kummer, dedekind, kummer dedekind, dedekind-kummer, dedekind kummer
+-/
+
+open_locale big_operators
+
+open ideal polynomial double_quot unique_factorization_monoid
+
+open_locale classical
+
+variables {R : Type*} [comm_ring R] [is_domain R] [is_dedekind_domain R]
+
+variables {S : Type*} [comm_ring S] [algebra R S]
+
+lemma temporary (f : polynomial R) (r : R) :
+  adjoin_root.of f r = adjoin_root.mk f (polynomial.C r) := rfl
+
+open alg_equiv
+
+/-- Let `f` be a polynomial over `R` and `I` an ideal of `R`,
+then `(R[x]/(f)) / (I)` is isomorphic to `(R/I)[x] / (f mod p)` -/
+noncomputable def adjoin_root.quot_equiv_quot_map
+  (f : polynomial R) (I : ideal R) :
+  (_ ⧸ (ideal.map (adjoin_root.of f) I)) ≃ₐ[R]
+    _ ⧸ (ideal.span ({polynomial.map I^.quotient.mk f} : set (polynomial (R ⧸ I)))) :=
+alg_equiv.of_ring_equiv (adjoin_root.quot_adjoin_root_equiv_quot_polynomial_quot I f)
+begin
+  intros x,
+  have : algebra_map R (adjoin_root f ⧸ (ideal.map (adjoin_root.of f) I)) x =
+    ideal.quotient.mk (ideal.map (adjoin_root.of f) I) (adjoin_root.of f x) := rfl,
+  rw temporary f x at this,
+  rw this,
+  rw adjoin_root.quot_adjoin_root_equiv_quot_polynomial_quot_mk_of,
+  have : algebra_map R (polynomial (R ⧸ I) ⧸ (ideal.span {polynomial.map
+    (ideal.quotient.mk I) f})) x =
+    ideal.quotient.mk (ideal.span {polynomial.map (ideal.quotient.mk I) f})
+    (polynomial.C (ideal.quotient.mk I x)) := rfl,
+  rw this,
+  simp only [map_C],
+end
+
+@[simp] lemma adjoin_root.quot_equiv_quot_map_apply
+  (f : polynomial R) (I : ideal R) (x : polynomial R) :
+  adjoin_root.quot_equiv_quot_map f I (ideal.quotient.mk _ (adjoin_root.mk f x)) =
+    ideal.quotient.mk _ (x.map I^.quotient.mk) :=
+by rw [adjoin_root.quot_equiv_quot_map, alg_equiv.of_ring_equiv_apply,
+    adjoin_root.quot_adjoin_root_equiv_quot_polynomial_quot_mk_of]
+
+lemma adjoin_root.quot_equiv_quot_map_symm_apply
+  (f : polynomial R) (I : ideal R) (x : polynomial R) :
```

Comment:
Please don't call a polynomial `x`... (and you can write `(f g : R[X])` if you open the locale polynomial).